cql3: time_uuid_fcts: validate time UUID

bhalevy opened the pull request:
Throw an error in case we hit an invalid time UUID
rather than hitting an assert.

Ref #5552

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

bhalevy commented:
In v2:
- added UUID.is_timestamp
- added asserts for create_time
- added unit tests for time uuid
- added validation for time UUID
- deleted outdated FIXME comment
- added time uuid validation tests
- abstract_function_selector: provide assignment_testable_source_context

Test: unit(dev), cql_cast_test:CQLCastTest.cast_timeuuid_test cql_additional_tests:TestCQL.timeuuid_test(dev)

psarna commented on cql3/functions/time_uuid_fcts.hh:
```diff
@@ -62,6 +62,18 @@ make_now_fct() {
+    int64_t ms = ts.time_since_epoch().count();
+    auto nanos_since = utils::UUID_gen::make_nanos_since(ms);
+    if (!utils::UUID_gen::is_valid_nanos_since(nanos_since)) {
+        if (!utils::UUID_gen::is_valid_nanos_since(nanos_since)) {
```
why is the same condition checked twice?

bhalevy replied:
hmm, this was unintentional.  Will fix.

bhalevy replied:
Fixed in v3

psarna commented:
I also see that there are unresolved conflicts with /next, so please rebase as well.

bhalevy commented:
In v3:
- deleted extraneous call to `is_valid_nanos_since`
- deleted inadvertent update to seastar submodule

**will rebase after next is promoted**

bhalevy commented:
Rebased onto a74a82d4d245d75311c19d0d8fba404d5d75937e and retested.

bhalevy commented:
ping @psarna

psarna commented:
Looks good, I hope that I didn't miss any cases when these asserts fire from user input :) I'll fetch these changes, retest, and queue.

bhalevy commented:
Dequeued due to cql_query_test failure in debug mode.

nyh commented:
Now you see why I don't like to mark the target of a pull request as "next" - your pull request got automatically closed as soon as it entered next, and now there is no way to reopen it, and you need to create a different one.

bhalevy commented:
> Now you see why I don't like to mark the target of a pull request as "next" - your pull request got automatically closed as soon as it entered next, and now there is no way to reopen it, and you need to create a different one.

True, that's an unfortunate outcome. Similarly, issues that get fixed by the pr are closed when it's merged to next, rather than when next is promoted to master. At least the issues can be reopened, unlike the pr.
